[Emily Ong] iP

EmilyOng opened the pull request:
# Duke

⚠️ Copied from https://nus-cs2103-ay2223s1.github.io/website/schedule/week4/project.html#1-use-gfmd-in-the-pr-description

> "Hello World!"

[Duke frees your mind of having to remember things you need to do](https://nus-cs2103-ay2223s1.github.io/website/schedule/week4/project.html#1-use-gfmd-in-the-pr-description).

**It's,**

- [x] Suited for fast typists
- [x] Has a gentle learning curve
- [x] Fast to use
- [x] Secure storage 101

**All you need to do is**

1. Download it from [here](https://github.com/EmilyOng/ip/releases).
2. Double-click it.
3. Add your tasks.
4. Let it manage your tasks for you 😉

**Features**

- Managing tasks, events & deadlines
- Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the main method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

EmilyOng committed:
Implement a skeletal version with Greet, Echo and Exit

EmilyOng committed:
Add clarifying comments and improve abstraction of output formatter

EmilyOng committed:
Add the ability to store text entered by the user and display them when requested

EmilyOng committed:
Rename command manager to task manager

EmilyOng committed:
Add the Task class to represent tasks

EmilyOng committed:
Add the ability to mark tasks as done or undone

EmilyOng committed:
Refactor printing to use DukePrinter class

EmilyOng committed:
Add clarifying comments and tidy up

EmilyOng committed:
Handle errors where task number provided may be invalid

EmilyOng committed:
Display message if task list is empty

EmilyOng committed:
Use inheritance to model various task types

EmilyOng committed:
Fix wrong input passed to task creation

EmilyOng committed:
Handle creation of ToDos, Events, Deadlines

EmilyOng committed:
Add a helper method for creating Task

EmilyOng committed:
Remove unused imports

EmilyOng committed:
Add simple tests using I/O redirections

EmilyOng committed:
Use switch statements for handling commands

EmilyOng committed:
Add DukeException class to handle Duke-specific errors

EmilyOng committed:
Implement delete functionality for task

EmilyOng committed:
Refactor project files into proper packages

EmilyOng committed:
Tidy up logics for task deletion and improve exception handling

EmilyOng committed:
Add enums for TaskType and CommandType

EmilyOng committed:
Remove unreadable emoticon

EmilyOng committed:
Support serialization for tasks

EmilyOng committed:
Centralize task serialization operations

EmilyOng committed:
Implement database facility with serializable objects

EmilyOng committed:
Remove tracking for data files

EmilyOng committed:
Add missing newline when appending to data file

EmilyOng committed:
Use LocalDate class for storing datetime objects in tasks

EmilyOng committed:
Improve error handling of invalid dates

EmilyOng committed:
Augment list command to filter tasks by date

EmilyOng committed:
Merge pull request #1 from EmilyOng/branch-Level-7

[Level-7] Save

EmilyOng committed:
Merge 'master' branch and fix merge conflicts

EmilyOng committed:
Merge pull request #2 from EmilyOng/branch-Level-8

[Level 8] Dates and Times

EmilyOng committed:
Fix tests

EmilyOng committed:
Restructure folders into domains

EmilyOng committed:
Implement Ui class

EmilyOng committed:
Implement Parser class and deprecate CommandManager

EmilyOng committed:
Handle status printing within commands

EmilyOng committed:
Improve packaging of classes

EmilyOng committed:
Tidy up and remove unused codes

EmilyOng committed:
Use the 'duke' package

EmilyOng committed:
Remove unused DukePrinter class

EmilyOng committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

EmilyOng committed:
Specify Duke application for Gradle

EmilyOng committed:
Move duke into its own package

EmilyOng committed:
Add unit test for ParserManager

EmilyOng committed:
Add unit tests for task models

* Add tests for mark as done and unmark as done methods
* Add tests for string conversion

EmilyOng committed:
Add missing return statements in List tasks

EmilyOng committed:
Fix mainClassName in Gradle build config

EmilyOng committed:
Implement find command to search for a task by keyword

* Search functionality checks for whether the provided keyword is
contained in the task's description

EmilyOng committed:
Set up checkstyle configurations

EmilyOng committed:
Reformat code to adhere to checkstyle specifications

EmilyOng committed:
Reformat code under test code

EmilyOng committed:
Add JavaDoc comments for the 'commands' package

EmilyOng committed:
Add JavaDoc comments for the 'exceptions' package

EmilyOng committed:
Add JavaDoc comments for the 'managers' package

EmilyOng committed:
Add JavaDoc comments for the 'models' package

EmilyOng committed:
Add JavaDoc comments for the 'storage' package

EmilyOng committed:
Add JavaDoc comments for the 'utils' package

EmilyOng committed:
Add JavaDoc comments for the main 'Duke' entry point

EmilyOng committed:
Fix typo

EmilyOng committed:
Serializable class: Add missing throws JavaDoc

EmilyOng committed:
Merge pull request #3 from EmilyOng/branch-Level-9

[Level 9] Find

EmilyOng committed:
Merge branch 'master' of https://github.com/EmilyOng/ip into A-JavaDoc

EmilyOng committed:
Merge pull request #5 from EmilyOng/A-JavaDoc

[A-JavaDoc] JavaDoc

EmilyOng committed:
Use a common prefix for associated constants

EmilyOng committed:
Merge 'master' branch and fix conflicts

EmilyOng committed:
Run reformat on code

EmilyOng committed:
Fix remnants of unformatted code

EmilyOng committed:
Merge pull request #6 from EmilyOng/branch-A-CodingStandard

[A-CodingStandard]: Follow the Coding Standard

EmilyOng committed:
Fix wrong reformat applied

EmilyOng committed:
Set up A-CheckStyle branch

EmilyOng committed:
Remove IO redirection tests

EmilyOng committed:
Add JFX dependency to Gradle build config

EmilyOng committed:
Refactor commands to return a status message

seelengxd commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,66 @@
+    /**
+     * The main event loop of the application.
+     *
+     * @param args Unused arguments
```
```suggestion
     * @param args Unused arguments.
```

Should there be punctuation added? [relevant coding standard](https://se-education.org/guides/conventions/java/intermediate.html#:~:text=Punctuation%20behind%20each%20parameter%20description)

EmilyOng replied:
Thanks!

seelengxd commented on src/main/java/duke/commands/ListTasksCommand.java:
```diff
@@ -0,0 +1,91 @@
+                            return taskDate != null && taskDate.isBefore(date);
+                        }
+                );
+                uiManager.print(TaskManager.display(filteredList));
```
```suggestion
                List<Task> filteredTasks = taskManager.list(
                        task -> {
                            LocalDate taskDate = task.getDate();
                            return taskDate != null && taskDate.isEqual(date);
                        }
                );
                uiManager.print(TaskManager.display(filteredTasks));
                return;
            }
            Matcher matchTasksBefore = ListTasksCommand.MATCH_TASKS_BEFORE.matcher(this.arguments);
            if (matchTasksBefore.matches()) {
                LocalDate date = DukeValidator.parseDate(matchTasksBefore.group("date"));
                List<Task> filteredTasks = taskManager.list(
                        task -> {
                            LocalDate taskDate = task.getDate();
                            return taskDate != null && taskDate.isBefore(date);
                        }
                );
                uiManager.print(TaskManager.display(filteredTasks));
```

Should this be a plural form? [relevant coding standard](https://se-education.org/guides/conventions/java/intermediate.html#:~:text=Plural%20form%20should%20be%20used%20on%20names%20representing%20a%20collection%20of%20objects.) ~~i am not too sure about this one tbh~~

seelengxd commented on src/main/java/duke/managers/ParserManager.java:
```diff
@@ -0,0 +1,72 @@
+        }
+        String command = matcher.group("commandWord").strip();
+        String arguments = matcher.group("arguments").strip();
+        switch (command) {
+        // List all tasks in the task manager
+        case ListTasksCommand.COMMAND_WORD:
+            return new ListTasksCommand(arguments);
+        case FindTaskCommand.COMMAND_WORD:
+            return new FindTaskCommand(arguments);
+
+        // Modifying operations on a task
+        case MarkTaskCommand.COMMAND_WORD:
+            return new MarkTaskCommand(arguments);
+        case UnmarkTaskCommand.COMMAND_WORD:
+            return new UnmarkTaskCommand(arguments);
+        case DeleteTaskCommand.COMMAND_WORD:
+            return new DeleteTaskCommand(arguments);
+
+        // Creating operations on a task
+        case AddToDoTaskCommand.COMMAND_WORD:
+            return new AddToDoTaskCommand(arguments);
+        case AddDeadlineTaskCommand.COMMAND_WORD:
+            return new AddDeadlineTaskCommand(arguments);
+        case AddEventTaskCommand.COMMAND_WORD:
+            return new AddEventTaskCommand(arguments);
+
+        // Termination of program
+        case ByeCommand.COMMAND_WORD:
+            return new ByeCommand();
+
+        default:
+            return new IncorrectCommand();
+        }
```
I like this part. It follows this [coding standard](https://se-education.org/guides/conventions/java/intermediate.html#layout:~:text=Logical%20units%20within%20a%20block%20should%20be%20separated%20by%20one%20blank%20line.). Logically separating parts of the case statement made this easy to read.

seelengxd commented on src/main/java/duke/commands/UnmarkTaskCommand.java:
```diff
@@ -0,0 +1,56 @@
+public class UnmarkTaskCommand implements Command {
+    public static final String COMMAND_WORD = "unmark";
+
+    private static final String MESSAGE_MARK_TASK_AS_UNDONE = "OK, I've marked this task as not done yet:";
+
+    private static final String ERROR_MISSING_TASK_INDEX = "You are missing a task number!\n"
+            + "Use the 'list' command to view the tasks and their number.";
+    private static final String ERROR_NAN_TASK_NUMBER = "The task number you provided is not a number!";
```
I liked how the commands generally have many constants (that follow the [coding standard](https://se-education.org/guides/conventions/java/intermediate.html#naming:~:text=Constant%20names%20must%20be%20all%20uppercase%20using%20underscore%20to%20separate%20words.)). The abstraction makes the code feel very clean.

seelengxd reviewed (commented):
Overall, I find your code quite impressive and readable. Aside from a few missing punctuations for javadoc, the coding standards seem to be followed. I also thought it was cool that there were so many packages and constants to organise your code. 😄

EmilyOng committed:
Implement GUI for chat

EmilyOng committed:
Remove unused UI manager

EmilyOng committed:
Add variable arguments to find command

EmilyOng committed:
Merge pull request #7 from EmilyOng/branch-A-CheckStyle

[A-CheckStyle]: Use CheckStyle

EmilyOng committed:
Merge pull request #8 from EmilyOng/branch-Level-10

[Level-10] GUI

* Implemented GUI using JavaFX with FXML definitions of views, found under `resources/views`
* Resultant unused classes including the UI Manager is now removed
* Tests are now conducted under JUnit and not I/O redirection

EmilyOng committed:
Merge 'master' branch and fix conflicts

EmilyOng committed:
Merge pull request #9 from EmilyOng/branch-A-Varargs

[A-Varargs] Use Varargs

* Varargs are implemented for the `find` command. Users can use `find <keyword1> <keyword2> ...` to perform a search on the list of tasks to retrieve the tasks matching all keywords.

EmilyOng committed:
Add Help command

EmilyOng committed:
Tweak padding of the dialog box

nareus commented on src/main/java/duke/managers/ParserManager.java:
```diff
@@ -0,0 +1,119 @@
+    public static Map<String, String> getAllCommands() {
+        return new LinkedHashMap<>() {
+            {
+                put(AddDeadlineTaskCommand.COMMAND_WORD, AddDeadlineTaskCommand.COMMAND_DESCRIPTION);
```
Could the name of the enumeration be shorter?

nareus commented on src/main/java/duke/managers/TaskManager.java:
```diff
@@ -0,0 +1,149 @@
+ * @author Emily Ong Hui Qi
+ */
+public class TaskManager {
+    private static final String MESSAGE_NO_TASKS_AVAILABLE = "There are currently no tasks available. Add one now!";
```
I like that you have followed the string format for the final attributes

nareus commented on src/main/java/duke/models/serializable/TaskSerializable.java:
```diff
@@ -0,0 +1,111 @@
+
+    private static final String ERROR_UNKNOWN_TASK_TYPE = "Unknown task type %s!";
+
+    private final TaskType taskType;
```
Appreciate the use of camel style for variables throughout the project.

EmilyOng replied:
Thanks!

nareus reviewed (commented):
Well done!

EmilyOng committed:
Add assertions to code

* Assertions are added to document important assumptions that should
hold at various points in the code.
* In this commit, assertions are added to ensure that the file storage
is not null when the programmer attempts to access the file storage.
Assertions are also added to ensure that tasks added are non-null and
the serializable format is correct.

EmilyOng committed:
Add missing full-stops to JavaDocs

* Coding standard: https://se-education.org/guides/conventions/java/intermediate.html#comments

EmilyOng committed:
Improve semantics of variable namings

EmilyOng committed:
ListTasksCommand: Make the happy path prominent

EmilyOng committed:
Avoid magic numbers

EmilyOng committed:
Add streams for displaying concatenated task display

EmilyOng committed:
Merge pull request #12 from EmilyOng/branch-A-Streams

[A-Streams] Use Streams

* Streams are used to display the list of tasks via the `list` command, which was previously accomplished with using string concatenation via `StringBuilder`.

EmilyOng committed:
Merge pull request #11 from EmilyOng/branch-A-CodeQuality

[A-CodeQuality] Improve Code Quality

* Addressed missing full-stops in JavaDocs via https://github.com/nus-cs2103-AY2223S1/ip/pull/24#discussion_r956847517
* Changed `filteredList` -> `filteredTasks` to improve readability via https://github.com/nus-cs2103-AY2223S1/ip/pull/24#discussion_r956854131
* Made the happy-path in `ListTasksCommand` prominent as suggested in https://nus-cs2103-ay2223s1.github.io/website/book/codeQuality/maximiseReadability/advanced/makeHappyPathProminent/
* Addressed magic numbers in UI display as suggested in https://nus-cs2103-ay2223s1.github.io/website/book/codeQuality/maximiseReadability/basic/avoidMagicNumbers/

In future, developers should refer to the [coding conventions](https://se-education.org/guides/conventions/java/intermediate.html) when writing code, particularly for implicit checks. Explicit checks can be done via the checkstyle tasks using Gradle.

EmilyOng committed:
Merge 'master' branch and fix conflicts

EmilyOng committed:
Merge pull request #10 from EmilyOng/branch-A-Assertions

[A-Assertions] Use Assertions

EmilyOng committed:
Show the number of tasks that have been completed in the past week

EmilyOng committed:
Merge pull request #13 from EmilyOng/branch-BCD-Extension

[BCD-Extension] Statistics and insights

EmilyOng committed:
Disallow sending message via enter key if message is empty

EmilyOng committed:
Set up GitHub actions

EmilyOng committed:
Add User Guide for Duke

EmilyOng committed:
Merge pull request #15 from EmilyOng/branch-A-UserGuide

[A-UserGuide] User Guide

EmilyOng committed:
Disable resizing of pane

EmilyOng committed:
Add Ui screenshot and set stage name

EmilyOng committed:
docs: Add release link and UI image
